src/lustre/lustreTypeChecker.ml

```diff
@@ -0,0 +1,1115 @@
+(* This file is part of the Kind 2 model checker.
+
+   Copyright (c) 2020 by the Board of Trustees of the University of Iowa
+
+   Licensed under the Apache License, Version 2.0 (the "License"); you
+   may not use this file except in compliance with the License.  You
+   may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0 
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+   implied. See the License for the specific language governing
+   permissions and limitations under the License. 
+
+ *)
+
+(* @author Apoorv Ingle *)
+
+(* TODO: Find strongly connected components, put them in a group *)
+(* TODO: This also checks if we have co-dependent types or values *)
+(* TODO: UserType if they are alias of any other type, then flatten them out *)
+(* TODO: Introduce GroupType which is like tuple type but has flattened cannonical structure*)
+
+module R = Res
+
+module LA = LustreAst
+module LC = LustreContext
+module LH = LustreAstHelpers
+
+(** Returns [ok] if the typecheck/typeinference runs fine 
+ * or reports an error at position with the error *)
+type 'a tcResult = ('a, Lib.position * string) result
+
+(** Type alias for lustre type from LustreAst  *)
+type tcType  = LA.lustre_type
+
+(** String of the type to display in type errors *)
+let string_of_tcType: tcType -> string = fun t -> Lib.string_of_t LA.pp_print_lustre_type t
+                       
+(** Map for types with identifiers as keys *)
+module IMap = struct
+  (** everything that [Stdlib.Map] gives us  *)
+  include Map.Make(struct
+              type t = LA.ident
+              let compare i1 i2 = Stdlib.compare i1 i2
+            end)
+  (** Pretty print type synonyms*)
+  let pp_print_type_binding ppf = fun i ty -> 
+    Format.fprintf ppf "(%a<->%a), " LA.pp_print_ident i LA.pp_print_lustre_type ty
+
+  (** Pretty print type bindings*)
+  let pp_print_type_binding ppf = fun i ty -> 
+    Format.fprintf ppf "(%a:%a), " LA.pp_print_ident i LA.pp_print_lustre_type ty
```

Comment:
These two functions have the same name. The second one overrides the first one.

Reply:
This is fixed now